### PR TITLE
added feature to evaluate simple arithmetic expressions

### PR DESCRIPTION
**Pull request by Majd223**

### CODE REVIEW

File: test/package.test.ts [hunk 1]
```diff
@@ -6,34 +7,104 @@ test("basic", () => {
...
+            variableValue: 5,
+        },
+        {
+            variableName: "w",
+            variableValue: 13,
         },
     ]);
 });
+
+// test("simple assignment", () => {
```
Comment (wandyezj) [inline test/package.test.ts]: How come these tests are commented out?

File: src/VariableStatementAnalysis.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
 export interface VariableStatementAnalysis {
-    text: string;
```
Comment (wandyezj) [inline src/VariableStatementAnalysis.ts:2]: How come `text` is removed?

`text` helps make the tests more readable.

File: src/analyzeCode.ts [hunk 3]
```diff
@@ -25,32 +30,149 @@ export function analyzeCode(code: string): VariableStatementAnalysis[] {
...
+    41: (a: number, b: number): number => a * b,
+    42: (a: number, b: number): number => a ** b,
+    43: (a: number, b: number): number => a / b,
+    44: (a: number, b: number): number => a % b,
+
+    //36: (a:number, b:number): boolean => a === b
+    //37: (a:number, b:number): boolean => a !== b
+};
+
+function getValue(
```
Comment (wandyezj) [inline src/analyzeCode.ts]: How come getting the value of `ts.NumericLiteral` and `ts.Identifier` are combined into one function?

File: test/package.test.ts [hunk 4]
```diff
@@ -6,34 +7,104 @@ test("basic", () => {
     expect(actual).toBe(expected);
 });
```
Comment (wandyezj) [inline test/package.test.ts:8]: Another test case `const x = - 1`.

File: src/analyzeCode.ts [hunk 5]
```diff
@@ -25,32 +30,149 @@ export function analyzeCode(code: string): VariableStatementAnalysis[] {
...
-                text,
+                variableName: node.name.getText(),
+
+                //TODO: can we reference detectedvariablestatements without passing it around?
+                variableValue: processExpression(
+                    node.initializer,
+                    detectedVariableStatements
+                ),
             });
+        } else if (ts.isExpressionStatement(node)) {
```
Comment (wandyezj) [inline src/analyzeCode.ts:69]: Should only the expression of the `VariableDeclaration` initializer be evaluated?

File: src/analyzeCode.ts [hunk 6]
```diff
@@ -25,32 +30,149 @@ export function analyzeCode(code: string): VariableStatementAnalysis[] {
...
-    // sourceFile.statements.forEach((node) => {
-    //     visitNodeRecursive(node, visitVariableStatement);
-    // });
 
     return detectedVariableStatements;
 }
 
+//TODO: can we merge from a record to something else that supports boolean 
+const Operations: Record<number, (a: number, b: number) => number> = {
+    39: (a: number, b: number): number => a + b,
```
Comment (wandyezj) [inline src/analyzeCode.ts]: Where do these magic numbers (example `39` ) come from? Is there an enum that can be used instead?

File: src/analyzeCode.ts [hunk 7]
```diff
@@ -25,32 +30,149 @@ export function analyzeCode(code: string): VariableStatementAnalysis[] {
...
+function processExpression(
+    node: ts.Expression | ts.NumericLiteral | undefined,
+    detectedVariableStatements: VariableStatementAnalysis[]
+): number {
+    if (node === undefined) {
+        throw new Error("Expression is undefined");
+    }
+
+    if (ts.isBinaryExpression(node)) {
+        if (
```
Comment (wandyezj) [inline src/analyzeCode.ts:246]: Could you solve this with recursion instead? To avoid the complex if?

`processExpression(node: ts.Expression | ts.NumericLiteral | ts.BinaryExpression): number | undefined`

note: in some cases the expression value may not be assessable, so undefined needs to be returned.

File: test/package.test.ts [hunk 8]
```diff
@@ -6,34 +7,104 @@ test("basic", () => {
...
+    expect(statements).toStrictEqual([
+        {
+            variableName: "y",
+            variableValue: 2,
+        },
+    ]);
+});
+
+test("Expression with multiple predefined variables", () => {
+    const code = `const x = 2 + 5 ;
```
Comment (wandyezj) [inline test/package.test.ts]: I find it helpful to use sequences of variable names if they are letters `a b c d e f` or  `x y z` and to only go out of sequence with a letter if there it is distinct in some way.

File: test/package.test.ts [hunk 9]
```diff
@@ -1,3 +1,4 @@
+import exp from "constants";
 import { analyzeCode, hello } from "../src/index";
```
Comment (wandyezj) [inline test/package.test.ts]: A test case for shadowing.

```typescript
let x = 1;
if (true) {
    let x = 2;
    let a = x;
}
```

What is the value of `a`?

File: src/analyzeCode.ts [hunk 10]
```diff
@@ -25,32 +30,149 @@ export function analyzeCode(code: string): VariableStatementAnalysis[] {
...
     function visitVariableStatement(node: ts.Node) {
-        if (node.kind === ts.SyntaxKind.VariableStatement) {
-            const text = node.getText();
+        //check if node is a binary expression
+        if (ts.isVariableDeclaration(node)) {
             detectedVariableStatements.push({
-                text,
+                variableName: node.name.getText(),
+
+                //TODO: can we reference detectedvariablestatements without passing it around?
```
Comment (wandyezj) [inline src/analyzeCode.ts]: per `can we reference detectedvariablestatements without passing it around?` not unless you put it in a class. It's ok to pass it around, there is a limited number of places it needs to be passed.

File: src/VariableStatementAnalysis.ts (hunk 2 above)
Comment (Majd223) [reply src/VariableStatementAnalysis.ts:2]: I started using this array to store variables with their values, you okay with that? or you want another array that stores something else beside the variables and their values?

File: test/package.test.ts (hunk 1 above)
Comment (Majd223) [reply test/package.test.ts]: Uncommented them

File: src/analyzeCode.ts (hunk 3 above)
Comment (Majd223) [reply src/analyzeCode.ts]: Moved them to the main processing function, makes it easier to process edge cases

File: src/analyzeCode.ts (hunk 7 above)
Comment (Majd223) [reply src/analyzeCode.ts:246]: I have cleaned up the code with recursion maybe give it a second look
but regarding the number | undefined... that will be problematic for the Record that maps to functions because you can't do undefined + 2 or something.... any solution for this?

File: test/package.test.ts (hunk 8 above)
Comment (Majd223) [reply test/package.test.ts]: Thanks for input, fixed with upcoming push

File: src/analyzeCode.ts (hunk 5 above)
Comment (Majd223) [reply src/analyzeCode.ts:69]: I don't think so? when you initialize a variable you get a variable statement that have a variableDeclaration which contains all the information that we need... but re-setting the same value later the node type is expression statement... so we should evaluate both types unless you are referring to something else

File: test/package.test.ts (hunk 9 above)
Comment (wandyezj) [reply test/package.test.ts]:
```typescript
const outer = 1;
const shadow = 1;
if (true) {
    const shadow = 2;
    const inner = 2;

    const test = shadow;
}

// cannot access inner
```

File: src/VariableStatementAnalysis.ts [hunk 11]
```diff
@@ -1,3 +1,4 @@
 export interface VariableStatementAnalysis {
```
Comment (wandyezj) [inline src/VariableStatementAnalysis.ts:1]: It can be helpful to get the line and starting character of the expression and variable identifier.

```typescript
node.pos
node.end

ts.getLineAndCharacterOfPosition
```

File: src/analyzeCode.ts (hunk 7 above)
Comment (Majd223) [reply src/analyzeCode.ts:246]: Do you know a way to throw an error and return undefined at the same time?

File: src/VariableStatementAnalysis.ts (hunk 2 above)
Comment (Majd223) [reply src/VariableStatementAnalysis.ts:2]: Text has been added back to variables
currently text doesn't get updated when resetting a variable, will fix that in the next push

File: test/package.test.ts (hunk 4 above)
Comment (Majd223) [reply test/package.test.ts:8]: PrefixUnaryExpression is working now, including --i or ++i
postfix coming soon

File: src/analyzeCode.ts (hunk 5 above)
Comment (Majd223) [reply src/analyzeCode.ts:69]: Now evaluating the variableDeclarationList, should add a feature to iterate and declare variables in that list

File: src/analyzeCode.ts (hunk 6 above)
Comment (Majd223) [reply src/analyzeCode.ts]: Moved everything to Map and numbers are now tokens that make sense and refer the typescript tokens enum

File: src/analyzeCode.ts (hunk 7 above)
Comment (Majd223) [reply src/analyzeCode.ts:246]: Function now returns undefined

File: test/package.test.ts (hunk 4 above)
Comment (Majd223) [reply test/package.test.ts:8]: postfix/prefix working now. can also throw an error when user writes --1 or 5++

File: src/VariableStatementAnalysis.ts (hunk 11 above)
Comment (Majd223) [reply src/VariableStatementAnalysis.ts:1]: Added

File: src/VariableStatementAnalysis.ts (hunk 2 above)
Comment (wandyezj) [reply src/VariableStatementAnalysis.ts:2]: There should be a separate entry for each assignment.

i.e.
```typescript

let x = 1; // Line A
x = 2; // Line B

```

Should have two separate entries. since on line A `x` has the value of 1 and on line B x has the value of 2.

**Review by wandyezj — APPROVED** [review]
